File: src/headless/certified-area-relaxation.py
#!/usr/bin/env python3
"""Certified Area Relaxation v1: mandatory rectangles, bounds, rotations, and NoOverlap2D only."""

import json
import math
import platform
import sys
import time


PROFILE = "certified-area-relaxation-v1"
OBJECTIVE = "horizontal-span-times-origin-anchored-height"


def emit(status, **fields):
    json.dump({
        "constraintProfile": PROFILE,
        "objective": OBJECTIVE,
        "status": status,
        "pythonVersion": platform.python_version(),
        **fields,
    }, sys.stdout, separators=(",", ":"))


try:
    import ortools
    from ortools.sat.python import cp_model
except ImportError:
    emit("dependency-missing")
    sys.exit(0)


def positive_integer(value, label):
    if isinstance(value, bool) or not isinstance(value, int) or value <= 0:
        raise ValueError(f"{label} must be a positive integer")
    return value


def parse_input():
    data = json.load(sys.stdin)
    if not isinstance(data, dict):
        raise ValueError("input must be an object")
    allowed_keys = {
        "constraintProfile", "objective", "devices", "limitWidth", "limitHeight",
        "allowRotate", "maxSeconds",
    }
    unexpected_keys = sorted(set(data) - allowed_keys)
    if unexpected_keys:
        raise ValueError(f"unsupported proof-model fields: {','.join(unexpected_keys)}")
    if data.get("constraintProfile") != PROFILE:
        raise ValueError("unexpected certified relaxation profile")
    if data.get("objective") != OBJECTIVE:
        raise ValueError("unexpected certified relaxation objective")
    devices = data.get("devices")
    if not isinstance(devices, list) or not devices:
        raise ValueError("devices must be a non-empty array")
    normalized_devices = []
    seen_ids = set()
    for index, device in enumerate(devices):
        if not isinstance(device, dict):
            raise ValueError(f"devices[{index}] must be an object")
        unexpected_device_keys = sorted(set(device) - {"id", "width", "height"})
        if unexpected_device_keys:
            raise ValueError(
                f"unsupported device proof fields: {','.join(unexpected_device_keys)}"
            )
        device_id = device.get("id")
        if not isinstance(device_id, str) or not device_id or device_id in seen_ids:
            raise ValueError(f"invalid or duplicate device ID at index {index}")
        seen_ids.add(device_id)
        normalized_devices.append({
            "id": device_id,
            "width": positive_integer(device.get("width"), f"{device_id}.width"),
            "height": positive_integer(device.get("height"), f"{device_id}.height"),
        })
    limit_width = positive_integer(data.get("limitWidth"), "limitWidth")
    limit_height = positive_integer(data.get("limitHeight"), "limitHeight")
    allow_rotate = data.get("allowRotate")
    if type(allow_rotate) is not bool:
        raise ValueError("allowRotate must be a boolean")
    max_seconds = data.get("maxSeconds")
    if isinstance(max_seconds, bool) or not isinstance(max_seconds, (int, float)) \
            or not math.isfinite(max_seconds) or max_seconds <= 0 or max_seconds > 30:
        raise ValueError("maxSeconds must be in (0, 30]")
    maximum_area = limit_width * limit_height
    if maximum_area >= 2 ** 62:
        raise ValueError("certified area domain exceeds the supported int64 range")
    return normalized_devices, limit_width, limit_height, allow_rotate, float(max_seconds)


def solve(devices, limit_width, limit_height, allow_rotate, max_seconds):
    model = cp_model.CpModel()
    variables = []
    x_intervals = []
    y_intervals = []
    for device in devices:
        device_id = device["id"]
        base_width = device["width"]
        base_height = device["height"]
        rotations = (0, 90, 180, 270) if allow_rotate else (0,)
        orientations = [
            (
                rotation,
                base_height if rotation in (90, 270) else base_width,
                base_width if rotation in (90, 270) else base_height,
            )
            for rotation in rotations
        ]
        rotation = model.new_int_var_from_domain(
            cp_model.Domain.from_values(rotations), f"rotation_{device_id}"
        )
        width = model.new_int_var(
            min(base_width, base_height), max(base_width, base_height), f"width_{device_id}"
        )
        height = model.new_int_var(
            min(base_width, base_height), max(base_width, base_height), f"height_{device_id}"
        )
        model.add_allowed_assignments([rotation, width, height], orientations)
        x = model.new_int_var(0, limit_width, f"x_{device_id}")
        y = model.new_int_var(0, limit_height, f"y_{device_id}")
        end_x = model.new_int_var(0, limit_width, f"end_x_{device_id}")
        end_y = model.new_int_var(0, limit_height, f"end_y_{device_id}")
        model.add(end_x == x + width)
        model.add(end_y == y + height)
        x_intervals.append(model.new_interval_var(x, width, end_x, f"x_interval_{device_id}"))
        y_intervals.append(model.new_interval_var(y, height, end_y, f"y_interval_{device_id}"))
        variables.append({"x": x, "end_x": end_x, "end_y": end_y})

    model.add_no_overlap_2d(x_intervals, y_intervals)
    minimum_x = model.new_int_var(0, limit_width, "minimum_x")
    maximum_x = model.new_int_var(0, limit_width, "maximum_x")
    maximum_y = model.new_int_var(0, limit_height, "maximum_y")
    model.add_min_equality(minimum_x, [entry["x"] for entry in variables])
    model.add_max_equality(maximum_x, [entry["end_x"] for entry in variables])
    model.add_max_equality(maximum_y, [entry["end_y"] for entry in variables])
    span_width = model.new_int_var(0, limit_width, "span_width")
    model.add(span_width == maximum_x - minimum_x)
    area = model.new_int_var(0, limit_width * limit_height, "area")
    model.add_multiplication_equality(area, [span_width, maximum_y])
    model.minimize(area)

    solver = cp_model.CpSolver()
    solver.parameters.max_time_in_seconds = max_seconds
    solver.parameters.num_search_workers = 1
    status = solver.solve(model)
    if status == cp_model.MODEL_INVALID:
        return {"status": "solver-failed"}
    if status == cp_model.INFEASIBLE:
        return {"status": "infeasible"}

    status_name = "optimal" if status == cp_model.OPTIMAL \
        else "feasible" if status == cp_model.FEASIBLE \
        else "unknown"
    response = solver.response_proto
    exact_lower_bound = int(response.inner_objective_lower_bound)
    raw_lower_bound = float(solver.best_objective_bound)
    if not math.isfinite(raw_lower_bound) \
            or abs(raw_lower_bound - exact_lower_bound) > 1e-6:
        raise RuntimeError("unit integer objective bound lost its exact representation")
    result = {
        "status": status_name,
        "rawBestObjectiveBound": raw_lower_bound,
        "certifiedIntegerLowerBound": exact_lower_bound,
    }
    if status in (cp_model.OPTIMAL, cp_model.FEASIBLE):
        incumbent = int(solver.value(area))
        if exact_lower_bound > incumbent:
            raise RuntimeError("certified lower bound exceeds the master incumbent")
        if status == cp_model.OPTIMAL and exact_lower_bound != incumbent:
            raise RuntimeError("optimal status did not close the master objective gap")
        result["masterIncumbentArea"] = incumbent
    return result


def main():
    started_at = time.monotonic()
    try:
        devices, limit_width, limit_height, allow_rotate, max_seconds = parse_input()
        result = solve(devices, limit_width, limit_height, allow_rotate, max_seconds)
        emit(
            result.pop("status"),
            orToolsVersion=ortools.__version__,
            elapsedMs=round((time.monotonic() - started_at) * 1000),
            **result,
        )
    except Exception as error:
        emit(
            "solver-failed",
            orToolsVersion=ortools.__version__,
            elapsedMs=round((time.monotonic() - started_at) * 1000),
            errorMessage=str(error)[:2000],
        )


if __name__ == "__main__":
    main()

File: src/headless/certified-area-relaxation.ts
import { spawnSync } from "node:child_process";
import { resolve } from "node:path";
import { fileURLToPath } from "node:url";

export const CERTIFIED_AREA_RELAXATION_PROFILE = "certified-area-relaxation-v1" as const;
export const CERTIFIED_AREA_RELAXATION_OBJECTIVE =
  "horizontal-span-times-origin-anchored-height" as const;

/** A globally mandatory, area-counted rectangle from the full problem. */
export interface CertifiedAreaRelaxationDevice {
  readonly id: string;
  readonly width: number;
  readonly height: number;
}

export type CertifiedAreaRelaxationStatus =
  | "optimal"
  | "feasible"
  | "infeasible"
  | "unknown"
  | "executable-missing"
  | "dependency-missing"
  | "timeout"
  | "solver-failed";

export interface CertifiedAreaRelaxationResult {
  readonly constraintProfile: typeof CERTIFIED_AREA_RELAXATION_PROFILE;
  readonly objective: typeof CERTIFIED_AREA_RELAXATION_OBJECTIVE;
  readonly status: CertifiedAreaRelaxationStatus;
  /** OR-Tools' floating representation of the best minimization lower bound. */
  readonly rawBestObjectiveBound?: number;
  /** Exact integer inner-objective bound for the unit-coefficient area IntVar. */
  readonly certifiedIntegerLowerBound?: number;
  /** A placement-only incumbent. It is not a routed upper bound. */
  readonly masterIncumbentArea?: number;
  readonly pythonVersion?: string;
  readonly orToolsVersion?: string;
  readonly elapsedMs?: number;
}

export interface CertifiedAreaRelaxationOptions {
  readonly devices: readonly CertifiedAreaRelaxationDevice[];
  readonly limitWidth: number;
  readonly limitHeight: number;
  readonly allowRotate: boolean;
  readonly maxSeconds: number;
  /** Test/embedding seam; production callers use the adjacent proof-only Python module. */
  readonly scriptPath?: string;
}

const FAILURE_PRIORITY: Readonly<Record<
  Extract<CertifiedAreaRelaxationStatus,
    "executable-missing" | "dependency-missing" | "timeout" | "solver-failed">,
  number
>> = {
  "executable-missing": 0,
  "dependency-missing": 1,
  "solver-failed": 2,
  timeout: 3,
};

/**
 * Solve the proof-only placement relaxation.
 *
 * The serialized DTO is deliberately reconstructed field-by-field. Search-only
 * properties attached by an untyped caller therefore cannot cross the proof
 * boundary into the Python model. The minimized relaxation is
 * (max(endX) - min(x)) * max(endY), so every full feasible layout maps to a
 * feasible master assignment whose objective does not exceed its charged area.
 */
export function solveCpSatAreaLowerBound(
  options: CertifiedAreaRelaxationOptions,
): CertifiedAreaRelaxationResult {
  validateOptions(options);
  if (options.devices.length === 0) {
    return {
      constraintProfile: CERTIFIED_AREA_RELAXATION_PROFILE,
      objective: CERTIFIED_AREA_RELAXATION_OBJECTIVE,
      status: "optimal",
      rawBestObjectiveBound: 0,
      certifiedIntegerLowerBound: 0,
      masterIncumbentArea: 0,
      elapsedMs: 0,
    };
  }

  const scriptPath = options.scriptPath ?? resolveProofScriptPath();
  const input = JSON.stringify({
    constraintProfile: CERTIFIED_AREA_RELAXATION_PROFILE,
    objective: CERTIFIED_AREA_RELAXATION_OBJECTIVE,
    devices: options.devices.map((device) => ({
      id: device.id,
      width: device.width,
      height: device.height,
    })),
    limitWidth: options.limitWidth,
    limitHeight: options.limitHeight,
    allowRotate: options.allowRotate,
    maxSeconds: options.maxSeconds,
  });
  const executables = [
    process.env["INDUSTRIAL_PLANNER_PYTHON"],
    "python3",
    "python",
  ].filter((value, index, all): value is string =>
    value !== undefined && value.length > 0 && all.indexOf(value) === index);

  let aggregateFailure: Extract<CertifiedAreaRelaxationStatus,
    "executable-missing" | "dependency-missing" | "timeout" | "solver-failed">
    = "executable-missing";
  let aggregatePythonVersion: string | undefined;
  let aggregateOrToolsVersion: string | undefined;
  for (const executable of executables) {
    const result = spawnSync(executable, [scriptPath], {
      input,
      encoding: "utf8",
      maxBuffer: 1024 * 1024,
      timeout: Math.ceil(options.maxSeconds * 1_000 + 15_000),
    });
    if (result.error !== undefined
      && (result.error as NodeJS.ErrnoException).code === "ENOENT") continue;
    const processErrorCode = (result.error as NodeJS.ErrnoException | undefined)?.code;
    if (result.signal !== null || processErrorCode === "ETIMEDOUT") {
      aggregateFailure = selectFailure(aggregateFailure, "timeout");
      continue;
    }
    if (result.stdout.trim().length === 0) {
      aggregateFailure = selectFailure(aggregateFailure, "solver-failed");
      continue;
    }

    try {
      const parsed = parseEnvelope(
        JSON.parse(result.stdout) as Record<string, unknown>,
        options.limitWidth * options.limitHeight,
      );
      aggregatePythonVersion = parsed.pythonVersion ?? aggregatePythonVersion;
      aggregateOrToolsVersion = parsed.orToolsVersion ?? aggregateOrToolsVersion;
      if (parsed.status === "optimal"
        || parsed.status === "feasible"
        || parsed.status === "infeasible"
        || parsed.status === "unknown") return parsed;
      aggregateFailure = selectFailure(aggregateFailure, parsed.status);
    } catch {
      aggregateFailure = selectFailure(aggregateFailure, "solver-failed");
    }
  }

  return {
    constraintProfile: CERTIFIED_AREA_RELAXATION_PROFILE,
    objective: CERTIFIED_AREA_RELAXATION_OBJECTIVE,
    status: aggregateFailure,
    ...(aggregatePythonVersion === undefined ? {} : { pythonVersion: aggregatePythonVersion }),
    ...(aggregateOrToolsVersion === undefined ? {} : { orToolsVersion: aggregateOrToolsVersion }),
  };
}

function resolveProofScriptPath(): string {
  try {
    return fileURLToPath(new URL("./certified-area-relaxation.py", import.meta.url));
  } catch {
    // Vitest/Vite may rewrite import.meta.url to a non-file module URL. The
    // headless command still runs from the package root in that environment.
    return resolve(process.cwd(), "src/headless/certified-area-relaxation.py");
  }
}

function validateOptions(options: CertifiedAreaRelaxationOptions): void {
  assertPositiveSafeInteger(options.limitWidth, "limitWidth");
  assertPositiveSafeInteger(options.limitHeight, "limitHeight");
  if (!Number.isSafeInteger(options.limitWidth * options.limitHeight)) {
    throw new Error("Certified area domain exceeds JavaScript's safe integer range");
  }
  if (typeof options.allowRotate !== "boolean") {
    throw new Error("allowRotate must be a boolean");
  }
  if (!Number.isFinite(options.maxSeconds) || options.maxSeconds <= 0 || options.maxSeconds > 30) {
    throw new Error(`maxSeconds must be in (0, 30], received ${options.maxSeconds}`);
  }
  const ids = new Set<string>();
  for (const device of options.devices) {
    if (typeof device.id !== "string" || device.id.length === 0) {
      throw new Error("Certified relaxation device IDs must be non-empty strings");
    }
    if (ids.has(device.id)) throw new Error(`Duplicate certified relaxation device ID: ${device.id}`);
    ids.add(device.id);
    assertPositiveSafeInteger(device.width, `${device.id}.width`);
    assertPositiveSafeInteger(device.height, `${device.id}.height`);
  }
}

function assertPositiveSafeInteger(value: number, label: string): void {
  if (!Number.isSafeInteger(value) || value <= 0) {
    throw new Error(`${label} must be a positive safe integer, received ${value}`);
  }
}

function parseEnvelope(
  value: Record<string, unknown>,
  maximumArea: number,
): CertifiedAreaRelaxationResult {
  if (value["constraintProfile"] !== CERTIFIED_AREA_RELAXATION_PROFILE) {
    throw new Error("Unexpected certified relaxation profile");
  }
  if (value["objective"] !== CERTIFIED_AREA_RELAXATION_OBJECTIVE) {
    throw new Error("Unexpected certified relaxation objective");
  }
  const status = value["status"];
  if (!isStatus(status)) throw new Error("Unexpected certified relaxation status");
  const pythonVersion = optionalString(value["pythonVersion"]);
  const orToolsVersion = optionalString(value["orToolsVersion"]);
  const elapsedMs = optionalNonnegativeNumber(value["elapsedMs"]);
  const base = {
    constraintProfile: CERTIFIED_AREA_RELAXATION_PROFILE,
    objective: CERTIFIED_AREA_RELAXATION_OBJECTIVE,
    status,
    ...(pythonVersion === undefined ? {} : { pythonVersion }),
    ...(orToolsVersion === undefined ? {} : { orToolsVersion }),
    ...(elapsedMs === undefined ? {} : { elapsedMs }),
  } as const;
  if (status === "infeasible"
    || status === "executable-missing"
    || status === "dependency-missing"
    || status === "timeout"
    || status === "solver-failed") return base;

  const rawBestObjectiveBound = optionalNonnegativeNumber(value["rawBestObjectiveBound"]);
  const certifiedIntegerLowerBound = optionalAreaInteger(
    value["certifiedIntegerLowerBound"], maximumArea,
  );
  const masterIncumbentArea = optionalAreaInteger(value["masterIncumbentArea"], maximumArea);
  if (rawBestObjectiveBound === undefined || certifiedIntegerLowerBound === undefined) {
    throw new Error("Solver did not return a certified objective bound");
  }
  if (Math.abs(rawBestObjectiveBound - certifiedIntegerLowerBound) > 1e-6) {
    throw new Error("Float and exact integer objective bounds disagree");
  }
  if ((status === "optimal" || status === "feasible") && masterIncumbentArea === undefined) {
    throw new Error("Solver status requires a master incumbent");
  }
  if (masterIncumbentArea !== undefined
    && certifiedIntegerLowerBound > masterIncumbentArea) {
    throw new Error("Certified lower bound exceeds the master incumbent");
  }
  if (status === "optimal" && certifiedIntegerLowerBound !== masterIncumbentArea) {
    throw new Error("Optimal relaxation must close its own objective gap");
  }
  return {
    ...base,
    rawBestObjectiveBound,
    certifiedIntegerLowerBound,
    ...(masterIncumbentArea === undefined ? {} : { masterIncumbentArea }),
  };
}

function isStatus(value: unknown): value is CertifiedAreaRelaxationStatus {
  return value === "optimal"
    || value === "feasible"
    || value === "infeasible"
    || value === "unknown"
    || value === "executable-missing"
    || value === "dependency-missing"
    || value === "timeout"
    || value === "solver-failed";
}

function optionalString(value: unknown): string | undefined {
  return typeof value === "string" && value.length > 0 ? value : undefined;
}

function optionalNonnegativeNumber(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) && value >= 0 ? value : undefined;
}

function optionalAreaInteger(value: unknown, maximumArea: number): number | undefined {
  return typeof value === "number"
    && Number.isSafeInteger(value)
    && value >= 0
    && value <= maximumArea
    ? value
    : undefined;
}

function selectFailure<T extends Extract<CertifiedAreaRelaxationStatus,
  "executable-missing" | "dependency-missing" | "timeout" | "solver-failed">>(
  left: T,
  right: T,
): T {
  return FAILURE_PRIORITY[right] > FAILURE_PRIORITY[left] ? right : left;
}
